### ISSUE TITLE
Allow cluster administrator to bootstrap Tenant namespaces

### ISSUE THREAD

**Issue by d-m** [clastix/capsule#174]
# Describe the feature

I've been working on an automated process using a Helm chart to provision tenants (the Tenant resource, as well as things like pod security policies and other cluster level resources). It would be helpful to be able to bootstrap namespaces for a tenant during this provisioning process.

Currently, when I, as a cluster admin, create a namespace named "my-tenant-namespace" for a tenant with the prefix "my-tenant", Capsule doesn't recognize the namespace as belonging to the tenant and doesn't create resource quotas, limits, etc. since the tenant user didn't create it.

It looks like this can be accomplished by adding the "ownerReferences" section to the Namespace manifest, but the UID of the Tenant resource is not easily determined via Helm:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: my-tenant-namespace
  ownerReferences:
  - apiVersion: capsule.clastix.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Tenant
    name: <tenant name>
    uid: <uid>
```

# What would the new user story look like?

1. Cluster administrator creates a namespace for a tenant
2. Cluster administrator annotates or labels the namespace with the Tenant resource name
2. Capsule controller configures Tenant ownership of the namespace based on annotation or label information

The existing `capsule.clastix.io/tenant` label could be used here.

# Expected behavior

After the cluster administrator labels the new namespace with the Tenant resource name, the capsule controller configures the Tenant owner as if it was created by a tenant admin.

**Comment by bsctl** [comment]
@d-m Thanks for suggesting this improvement. We're always eager of feedbacks from users.

Actually, you can achieve the same as cluster admin, by impersonate the tenant owner, `alice` in this case:

```
$ kubectl create ns oil-dm3 --as alice --as-group capsule.clastix.io --v=8
I1221 23:18:34.393103   53039 loader.go:375] Config loaded from file:  /Users/adriano/.kube/cmp
I1221 23:18:34.401291   53039 request.go:1068] Request Body: {"apiVersion":"v1","kind":"Namespace","metadata":{"creationTimestamp":null,"name":"oil-dm3"},"spec":{},"status":{}}
I1221 23:18:34.401403   53039 round_trippers.go:420] POST https://cmp.clastix.io:6443/api/v1/namespaces
I1221 23:18:34.401414   53039 round_trippers.go:427] Request Headers:
I1221 23:18:34.401419   53039 round_trippers.go:431]     Accept: application/json
I1221 23:18:34.401424   53039 round_trippers.go:431]     Impersonate-User: alice
I1221 23:18:34.401428   53039 round_trippers.go:431]     Impersonate-Group: capsule.clastix.io
I1221 23:18:34.401439   53039 round_trippers.go:431]     User-Agent: kubectl/v1.18.5 (darwin/amd64) kubernetes/e6503f8
I1221 23:18:35.209543   53039 round_trippers.go:446] Response Status: 201 Created in 808 milliseconds
I1221 23:18:35.209567   53039 round_trippers.go:449] Response Headers:
I1221 23:18:35.209576   53039 round_trippers.go:452]     Date: Mon, 21 Dec 2020 22:18:35 GMT
I1221 23:18:35.209582   53039 round_trippers.go:452]     Cache-Control: no-cache, private
I1221 23:18:35.209588   53039 round_trippers.go:452]     Content-Type: application/json
I1221 23:18:35.209594   53039 round_trippers.go:452]     Content-Length: 646
I1221 23:18:35.209633   53039 request.go:1068] Response Body: {"kind":"Namespace","apiVersion":"v1","metadata":{"name":"oil-dm3","selfLink":"/api/v1/namespaces/oil-dm3","uid":"4ef6a318-7208-41d1-85d7-ad4cd995d4fa","resourceVersion":"22327776","creationTimestamp":"2020-12-21T22:18:35Z","ownerReferences":[{"apiVersion":"capsule.clastix.io/v1alpha1","kind":"Tenant","name":"oil","uid":"1e987fba-7b91-4b1b-92ff-0aa7dd83f199","controller":true,"blockOwnerDeletion":true}],"managedFields":[{"manager":"kubectl","operation":"Update","apiVersion":"v1","time":"2020-12-21T22:18:35Z","fieldsType":"FieldsV1","fieldsV1":{"f:status":{"f:phase":{}}}}]},"spec":{"finalizers":["kubernetes"]},"status":{"phase":"Active"}}
namespace/oil-dm3 created

$ kubectl describe ns oil-dm3
Name:         oil-dm3
Labels:       capsule.clastix.io/tenant=oil
```

**Comment by bsctl** [comment]
With Helm, you can use 

```
--kube-as-group stringArray   Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
--kube-as-user string         Username to impersonate for the operation
```

**Comment by d-m** [comment]
Thanks for the reply. This would work great for Helm charts consisting solely of resources that a tenant admin has permission to deploy. However, this helm chart also provisions cluster level resources like the Tenant resource itself as well as PodSecurityPolicy resources.

**Comment by bsctl** [comment]
Interesting case. We may implement on the tenant controller a sort of `namespaceSelector` based on the label `capsule.clastix.io/tenant=<tenant_name>`  as the replicaSet does with pods. We need to evaluate technical impacts.

**Comment by prometherion** [comment]
What about using Helm hooks to issue a simple `kubectl` create?

```
kubectl create namespace my-namespace --dry-run=client -o yaml | kubectl apply --as=alice --as-group=capsule.clastix.io --v=10 -f -
```

This should be idempotent, so creating Namespace, or doing nothing if already there.

**Comment by d-m** [comment]
It ended up being a little more complicated. Using the `--as` and `--as-group` flags in a pod clears out the current context, which required providing kubectl with the server, CA cert, port, and token. The pod's service account needed permission to impersonate the tenant name and capsule group as well.

I got it to work with the following job, in case anyone else is interested:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ template "tenant.fullname" . }}-admin
  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-delete-policy": hook-succeeded
  labels:
    {{- include "tenant.labels" . | nindent 4 }}
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: {{ template "tenant.fullname" . }}-admin-impersonator
  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-delete-policy": hook-succeeded
  labels:
    {{- include "tenant.labels" . | nindent 4 }}
rules:
- apiGroups: [""]
  resources: ["users"]
  verbs: ["impersonate"]
  resourceNames: ["{{ include "tenant.tenantName" $ }}-admin"]
- apiGroups: [""]
  resources: ["groups"]
  verbs: ["impersonate"]
  resourceNames: ["clastix.capsule.io"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: {{ template "tenant.fullname" . }}-admin-impersonate
  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-delete-policy": hook-succeeded
  labels:
    {{- include "tenant.labels" . | nindent 4 }}
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: {{ template "tenant.fullname" . }}-admin-impersonator
subjects:
- kind: ServiceAccount
  name: {{ template "tenant.fullname" . }}-admin
  namespace: default
---
apiVersion: batch/v1
kind: Job
metadata:
  name: {{ template "tenant.fullname" . }}-bootstrap-namespaces
  annotations:
    "helm.sh/hook": post-install,post-upgrade
    "helm.sh/hook-delete-policy": hook-succeeded
  labels:
    {{- include "tenant.labels" . | nindent 4 }}
spec:
  template:
    metadata:
      name: {{ template "tenant.fullname" . }}-bootstrap-namespaces
      labels:
        {{- include "tenant.labels" . | nindent 8 }}
    spec:
      serviceAccountName: {{ template "tenant.fullname" . }}-admin
      containers:
        - name: kubectl
          image: "k8s.gcr.io/hyperkube:v1.12.1"
          imagePullPolicy: "IfNotPresent"
          command:
          - /bin/sh
          - -c
          - >
            kubectl create ns my-tenant-namespace --as=my-tenant-admin --as-group=clastix.capsule.io -s https://kubernetes.default --certificate-authority=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt --token=`cat /var/run/secrets/kubernetes.io/serviceaccount/token` -v=10
      restartPolicy: OnFailure
```

**Comment by prometherion** [comment]
Thanks for sharing this with the whole community, @d-m!

Just a little remark: instead of using `hyperkub`e image that is heavy a a really heavy thing you could use use the `bitnami/kubectl` one, and maybe using the Capsule service account in order to avoid installing additional resources that could be hard to follow.

Anyway, I'm really pleases to see your suggestion in how increase the use cases for Capsule: would you like to join our Slack workspace?

Reach out me or @bsctl as you prefer in order to send you an invite!

**Comment by d-m** [comment]
Actually, it looks like this raised another issue when I changed the command to use the create | apply pattern above to avoid an error when the namespace already exists.

This approach results in the following error for existing namespaces or creating new namespaces:

```bash
kubectl create ns my-tenant-namespace --dry-run=true -o yaml | kubectl apply -f -                                                                                                                                                                
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "/v1, Resource=namespaces", GroupVersionKind: "/v1, Kind=Namespace"
Name: "my-tenant-namespace", Namespace: ""
from server for: "STDIN": namespaces "my-tenant-namespace" is forbidden: User "my-tenant-admin" cannot get resource "namespaces" in API group "" in the namespace "my-tenant-namespace"
```

Running `kubectl create ns my-tenant-namespace` works as expected.

**Comment by d-m** [comment]
> Thanks for sharing this with the whole community, @d-m!
> 
> Just a little remark: instead of using `hyperkub`e image that is heavy a a really heavy thing you could use use the `bitnami/kubectl` one, and maybe using the Capsule service account in order to avoid installing additional resources that could be hard to follow.
> 
> Anyway, I'm really pleases to see your suggestion in how increase the use cases for Capsule: would you like to join our Slack workspace?
> 
> Reach out me or @bsctl as you prefer in order to send you an invite!

@prometherion Good idea about switching away from the hypercube image. Also, happy to join your Slack workspace; I'll DM  you at the Twitter profile you have in your GitHub profile.
